src/context/provided_types.rs

```diff
@@ -0,0 +1,92 @@
+use std::{marker::PhantomData, mem, sync::OnceLock};
+
+use better_any::TidAble;
+
+use crate::FileRunContext;
+
+pub trait FromFileRunContextInstanceProvider<'a>: Sized {
+    type Parent: FromFileRunContextInstanceProviderFactory<Provider<'a> = Self>;
+
+    fn get<T: FromFileRunContext<'a> + TidAble<'a>>(
+        &self,
+        file_run_context: FileRunContext<'a, '_, Self::Parent>,
+    ) -> Option<&T>;
+}
+
+pub trait FromFileRunContextInstanceProviderFactory: Send + Sync {
+    type Provider<'a>: FromFileRunContextInstanceProvider<'a, Parent = Self>;
+
+    fn create<'a>(&self) -> Self::Provider<'a>;
+}
+
+pub trait FromFileRunContext<'a> {
+    fn from_file_run_context(
+        file_run_context: FileRunContext<'a, '_, impl FromFileRunContextInstanceProviderFactory>,
+    ) -> Self;
+}
+
+mod _sealed {
+    use super::*;
+
+    pub trait Sealed<'a> {}
+
+    impl<'a, T1: FromFileRunContext<'a>> Sealed<'a> for (T1,) {}
+}
+
+pub trait FromFileRunContextProvidedTypes<'a>: _sealed::Sealed<'a> {
+    type OnceLockStorage: FromFileRunContextProvidedTypesOnceLockStorage<'a>;
+
+    // fn len(&self) -> usize;
+    // fn get_index<'a, T: FromFileRunContext<'a>>(&self) -> Option<usize>;
```

Comment:
Thought I was going to need something trickier for "mapping" into the individual plugins' provided types/"storage" but no doesn't look like it so I can remove these